lib/spark.js

```diff
@@ -395,3 +443,65 @@ function mapErrorToStatusCode(err) {
   // Fallback code for unknown errors
   return 600
 }
+
+/**
+ * Assigns weights to providers based on their protocol and context ID and picks one at random.
+ * Providers with higher weights have a higher chance of being selected.
+ *
+ * Providers serving content using Bitswap protocol are filtered out.
+ *
+ * @param {Provider[]} providers
+ * @returns {Provider | undefined}
+ */
+export function pickRandomProvider(providers) {
+  const filteredProviders = providers.filter((provider) => provider.protocol !== 'bitswap')
+  if (!filteredProviders.length) return
+  const weightedProviders = weighProviders(filteredProviders)
+  return pickRandomWeightedItem(weightedProviders)
+}
+
+/**
+ * Assigns weights to providers based on their protocol and context ID.
+ *
+ * HTTP providers and those whose context ID starts with 'gHa' are given higher weights,
+ * hence having a higher chance of being selected.
+ *
+ * @param {Provider[]} providers
+ * @returns {Provider & { weight: number }[]}
+ */
+function weighProviders(providers) {
```

Comment:
nit: This function would be cleaner as singular `weighProvider`, the iteration isn't part of its responsibility